### PR TITLE
T18936 include list of test-plan as sub-view for soc

### PR DESCRIPTION
**Pull request by alexandrasp**
Add a view showing all the test plans with the total number of test cases for
a particular SoC type.

Signed-off-by: Alexandra Pereira <alexandra.pereira@collabora.com>

### CODE REVIEW

**Comment by alexandrasp** [comment]
@gctucker I think that there is a problem in counts on this PR(test count), I am checking this, I will see how it works on staging. If you have any idea about the problem, let me know...

**Comment by gctucker** [comment]
@alexandrasp There's a problem in the commit summary: `include list of test-plan assub-view`

**Comment by gctucker** [comment]
@alexandrasp Also there's a Javascript error so it's failing to start the frontend service:
```
Error: Error: Parse error using esprima for file: /tmp/assets-build/js/app/view-socs-soc-job-kernel.2017.4.js
Error: Line 603: Unexpected token )
```

**Review by gctucker — CHANGES_REQUESTED** [review]
Please fix Javascript error and clarify git commit summary.

**Comment by gctucker** [comment]
Tested on staging: https://staging.kernelci.org/soc/rockchip/job/kernelci/kernel/staging-20200213.0/

**Comment by gctucker** [comment]
There are still several issues wit this view:
* The pie chart is showing a number of "boots" rather than "tests".
* The total numbers don't match what is expected, showing 5/0/0 instead of 468/5/442 on the example from the link above.
* This view should not have the list of expandable panels "Boards Tested" section but only the "Available Test Plans" to be like the Tests views.  The next view would show "Boards Tested" when selecting a single test plan.
* The numbers in the "Available Test Plans" don't seem correct, they should add up to 468/5/442.  It looks like `mach` is not used in the filters and all the results for that kernel are being counted.

**Review by gctucker — CHANGES_REQUESTED** [review]
Please fix the issues as per my previous comment.

**Comment by alexandrasp** [comment]
> There are still several issues wit this view:
> 
>     * The pie chart is showing a number of "boots" rather than "tests".
> 
>     * The total numbers don't match what is expected, showing 5/0/0 instead of 468/5/442 on the example from the link above.
> 
>     * This view should not have the list of expandable panels "Boards Tested" section but only the "Available Test Plans" to be like the Tests views.  The next view would show "Boards Tested" when selecting a single test plan.
> 
>     * The numbers in the "Available Test Plans" don't seem correct, they should add up to 468/5/442.  It looks like `mach` is not used in the filters and all the results for that kernel are being counted.

I suppose all those were fixed ;)

**Comment by alexandrasp** [comment]
@gctucker seems that is not in staging yet.

**Comment by gctucker** [comment]
> @gctucker seems that is not in staging yet.

@alexandrasp It is now :)

**Comment by alexandrasp** [comment]
@gctucker I detected a problem on pie-chart count, fixing it! Let me know if you find anything else..

**Comment by alexandrasp** [comment]
@gctucker could you update staging? Padovan suggested changing the title and make git url as link and I considered good points. What do you think?

File: app/dashboard/static/css/bootstrap-3.3.6.min.css [hunk 1]
```diff
@@ -3,4 +3,4 @@
...
...elative}.input-group-btn>.btn+.btn{margin-left:-1px}.input-group-btn>.btn:active,.input-group-btn>.btn:focus,.input-group-btn>.btn:hover{z-index:2}.input-group-btn:first-child>.btn,.input-group-btn:first-child>.btn-group{margin-right:-1px}.input-group-btn:last-child>.btn,.input-group-btn:last-child>.btn-group{z-index:2;margin-left:-1px}.nav{padding-left:0;margin-bottom:0;list-style:none}.nav>li{position:relative;display:block}.nav>li>a{position:relative;display:block;padding:10px 15px}.nav>li>a:focus,.nav>li>a:hover{text-decoration:none;background-color:#eee}.nav>li.disabled>a{color:#777}.nav>li.disabled>a:focus,.nav>li.disabled>a:hover{color:#777;text-decoration:none;cursor:not-allowed;background-color:transparent}.nav .open>a,.nav .open>a:focus,.nav .open>a:hover{background-color:#eee;border-color:#337ab7}.nav .nav-divider{height:1px;margin:9px 0;overflow:hidden;background-color:#e5e5e5}.nav>li>a>img{max-width:none}.nav-tabs{border-bottom:1px solid #ddd}.nav-tabs>li{float:left;margin-bottom:-1px}.nav-tabs>li>a{margin-right:2px;line-height:1.42857143;border:1px solid transparent;border-radius:4px 4px 0 0}.nav-tabs>li>a:hover{border-color:#eee #eee #ddd}.nav-tabs>li.active>a,.nav-t
```
Comment (gctucker) [inline app/dashboard/static/css/bootstrap-3.3.6.min.css]: Why is this change needed?

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 2]
```diff
@@ -20,381 +23,45 @@
...
-    'utils/storage',
-    'utils/session',
     'charts/passpie',
-    'components/boot/unique',
-    'utils/filter',
-    'components/boot/view',
+    'utils/table',
+    'tables/test',
     'utils/date',
     'sprintf'
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: These 2 are not used as far as I can tell, in which case they can be dropped from the list.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 3]
```diff
@@ -20,381 +23,45 @@
...
-                .fail(error.error, uniqueCountFail)
-                .done(uniqueCountDone);
-        } else {
-            html.replaceByClassTxt('unique-values', '?');
-        }
-    }
-
     function getBootsFailed() {
         html.replaceContent(
             document.getElementById('accordion-container'),
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: I don't think there are any `accordion-container` elements any more in this view.  Can you locally get this function called instead of `getBootsDone` to make it show the error message?

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 4]
```diff
@@ -20,381 +23,45 @@
...
-
-            $.when(deferred, unique.countD(response))
-                .fail(error.error, uniqueCountFail)
-                .done(uniqueCountDone);
-        } else {
-            html.replaceByClassTxt('unique-values', '?');
-        }
-    }
-
     function getBootsFailed() {
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: It's not boots any more, we could call it `plans` instead i.e. `getPlansFailed`.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 5]
```diff
@@ -20,381 +23,45 @@
...
-                } else {
-                    html.addClass(
-                        document.getElementById('all-btn'), 'active');
-                }
-            }
-        }
-    }
-
     function updateDetails(response) {
+        results = response.result;
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: This variable is now set before it is defined.  I think the original line below should be left where it is.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 6]
```diff
@@ -545,14 +204,6 @@ define([
...
 
-    function getBootDoneChart(response) {
-        chart.bootpie({
-            element: 'boot-chart',
-            response: response,
-            legend: true
-        });
-    }
-
     function getBoots() {
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: Rename to `getPlans`.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 7]
```diff
@@ -591,83 +230,345 @@ define([
...
-            gSessionStorage.addObjects(pageState).save();
+        function createBatchOp(result) {
+            var job = result.job;
+            var kernel = result.kernel;
+            var branch = result.git_branch;
+            var plan = result.name;
+            var mach = gSoc;
+            var qStr;
+
+            qStr = 'job=' + job;
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: Please use `URI.buildQuery` as in this example:
https://github.com/kernelci/kernelci-frontend/blob/master/app/dashboard/static/js/app/view-tests-job-branch-kernel-plan.2020.1.js#L362

It's a much safer way of constructing the query because it will encode special characters.  I hit an issue with the simple approach of creating a plain string when defconfig values had `+` and `=` characters in them.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 8]
```diff
@@ -591,83 +230,345 @@ define([
...
+        function createBatchOp(result) {
+            
+            var job = result.job;
+            var kernel = result.kernel;
+            var branch = result.git_branch;
+            var plan = result.name;
+            var mach = gSoc;
+            var qStr;
+
+            qStr = 'job=' + job;
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: Same comment about `URI.buildQuery`.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 9]
```diff
@@ -545,14 +204,6 @@ define([
...
-    function getBootDoneChart(response) {
-        chart.bootpie({
-            element: 'boot-chart',
-            response: response,
-            legend: true
-        });
-    }
-
     function getBoots() {
         var deferred;
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js:158]: These boot results are not needed, and in fact they're not used anywhere.  You should just have to call `getPlans`.

**Review by gctucker — CHANGES_REQUESTED** [review]
The test numbers look fine now on this view:

    https://staging.kernelci.org/soc/rockchip/job/kernelci/kernel/staging-20200213.0/

However some code about getting boot results needs to be removed, it seems to be a bit mixed up between the original file and the test results view.  There are also some minor things to clean-up.

**Comment by gctucker** [comment]
@alexandrasp Staging is updated at least once a day, no need to request it each time.  It's not fully automated yet but it will be soon.

File: app/dashboard/static/css/bootstrap-3.3.6.min.css (hunk 1 above)
Comment (alexandrasp) [reply app/dashboard/static/css/bootstrap-3.3.6.min.css]: @gctucker 'Cause it was showing a warning message on console, this file was trying to get the map with a wrong name.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js (hunk 7 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: updated.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js (hunk 8 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: also updated.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js (hunk 3 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: This method is not necessary anymore, I removed it! Now the code is using same logic from job page, as you can see here: https://github.com/alexandrasp/kernelci-frontend/blob/t18936-add-soc-sub-view-test-plan/app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js#L429

File: app/dashboard/static/css/bootstrap-3.3.6.min.css (hunk 1 above)
Comment (gctucker) [reply app/dashboard/static/css/bootstrap-3.3.6.min.css]: OK, that's a bit surprising as it has been used for a long time in other places.  I'll try to reproduce this while I'm working on the frontend.

File: app/dashboard/static/js/app/view-socs-soc-job.2017.3.3.js [hunk 10]
```diff
@@ -244,8 +249,6 @@ require([
     function getTestsDone(response) {
         var columns,
             results;
-
-        console.log(response)
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job.2017.3.3.js:248]: Erm, yes, sorry :angel:

**Review by gctucker — COMMENTED** [review]
Please remove the "boot reports" links as these are going to be removed.
https://github.com/kernelci/kernelci-frontend/pull/70/files#diff-e4600ee66b5546cf4cba44d827b5be6aR100-R106
https://github.com/kernelci/kernelci-frontend/pull/70/files#diff-e4600ee66b5546cf4cba44d827b5be6aR133-R141

**Comment by gctucker** [comment]
> Please remove the "boot reports" links as these are going to be removed.
> https://github.com/kernelci/kernelci-frontend/pull/70/files#diff-e4600ee66b5546cf4cba44d827b5be6aR100-R106
> https://github.com/kernelci/kernelci-frontend/pull/70/files#diff-e4600ee66b5546cf4cba44d827b5be6aR133-R141

I mean, the boot views are going to be removed, so we shouldn't be adding links to them now.

We may have some test views to link later on, but we can add them when we have them.

**Comment by alexandrasp** [comment]
> the boot views are going to be removed, so we shouldn't be adding links to them now.

I see and code is updated :)

**Comment by gctucker** [comment]
Please make each row in the table link to a test view with the SoC name in the query, for example:
https://staging.kernelci.org/test/job/kernelci/branch/staging.kernelci.org/kernel/staging-20200221.0/plan/baseline/?qemu

This should then appear in the search box like it does with the jobs page:
https://kernelci.org/job/?mainline

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 11]
```diff
@@ -20,380 +23,36 @@
...
-                    );
-                } else {
-                    html.addClass(
-                        document.getElementById('all-btn'), 'active');
-                }
-            }
-        }
-    }
-
     function updateDetails(response) {
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: I think the argument `response` should be renamed `results` to avoid ambiguity.  The results are taken from the response.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 12]
```diff
@@ -402,19 +61,15 @@ define([
             gitCommit,
             gitURL,
             gitURLs,
-            results,
             tooltipNode;
 
-        results = response.result;
-        if (results.length === 0) {
+        if (response.length === 0) {
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: This is still needed?  The function that calls `updateDetails` is already checking for the length of results and this variable is in fact `response.results[0]`.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 13]
```diff
@@ -545,129 +169,340 @@ define([
...
+
+        response.result.forEach(parseBatchData)
+    }
+
+    function getBatchStatus(results) {
+        var batchOps;
+        var deferred;
+
+        function createBatchOp(result) {
+
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: Superfluous blank line.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 14]
```diff
@@ -545,129 +169,340 @@ define([
...
+        plansFailed();
+    }
+
+    function getPlansDone(response) {
+        if (response.result.length === 0) {
+            getPlansFailed();
+            return
+        }
+    
+        updateDetails(response.result[0]);
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js:411]: See my comment below, about the value passed to `updateDetails`.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 15]
```diff
@@ -701,10 +536,15 @@ define([
...
-    getBoots();
-    registerEvents();
+
+    gPlansTable = table({
+        tableId: 'planstable',
+        tableLoadingDivId: 'table-loading',
+        tableDivId: 'table-div',
+    });
+
+    getPlans();
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: Maybe add timeouts like on other views?  Is there any good reason to not have timeouts here?

**Review by gctucker — COMMENTED** [review]
A few more minor things to tweak.

File: app/dashboard/static/css/bootstrap-3.3.6.min.css (hunk 1 above)
Comment (gctucker) [reply app/dashboard/static/css/bootstrap-3.3.6.min.css]: @alexandrasp OK I guess this is because it was probably generated using a symlink originally.  Did you see the issue because you tried to regenerate it?  I mean, this is only a comment so it shouldn't cause any browser errors.

In any case, this is not related to the SoC view changes.  So please move this to a separate PR if it's really needed.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 16]
```diff
@@ -701,10 +533,15 @@ define([
...
+        tableId: 'planstable',
+        tableLoadingDivId: 'table-loading',
+        tableDivId: 'table-div',
+    });
+
+    setTimeout(getPlans, 10);
+    setTimeout(getTestCount, 10);
 
     init.hotkeys();
     init.tooltip();
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: And, how about `setTimeout` for these functions?  They have timeouts in other views.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 17]
```diff
@@ -20,654 +23,483 @@
...
+                orderable: false,
+                className: 'pull-center',
+                render: _renderPlanStatus,
+            },
+        ]
+        gPlansTable
+            .data(results)
+            .columns(columns)
+            .order([0, 'asc'])
+            .rowURL('/soc/%(soc)s/job/%(job)s/kernel/%(kernel)s/plan/%(name)s/')
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: The `soc` value is undefined, causing URLs like this one:
```
https://staging.kernelci.org/soc/null/job/kernelci/kernel/staging-20200224.2/plan/baseline/
```
with `/soc/null/`.  Maybe use `gSoC`, or `mach` from the query data?

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js [hunk 18]
```diff
@@ -20,654 +23,483 @@
...
+            parent_id: 'null',
+            sort: 'name',
+            field: [
+                'name',
+                'created_on',
+                'job',
+                'git_branch',
+                'git_commit',
+                'git_url',
+                'kernel',
```
Comment (gctucker) [inline app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js:437]: Maybe add `mach` here to have it in `updatePlansTable()` and fix the URLs?

**Review by gctucker — COMMENTED** [review]
One more small thing about URLs from the table rows.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js (hunk 16 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: good point, updated.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js (hunk 17 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js]: also updated.

File: app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js (hunk 18 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-socs-soc-job-kernel.2017.4.js:437]: updated

**Comment by alexandrasp** [comment]
@gctucker is there any link to this new page?

File: app/dashboard/static/css/bootstrap-3.3.6.min.css (hunk 1 above)
Comment (alexandrasp) [reply app/dashboard/static/css/bootstrap-3.3.6.min.css]: sure, already removed from this PR

**Review by gctucker — APPROVED** [review]
Tested OK on staging.

**Comment by gctucker** [comment]
Example view on staging:
https://staging.kernelci.org/soc/rockchip/job/kernelci/kernel/staging-20200213.0/
